torch/_custom_op.py

```diff
@@ -0,0 +1,292 @@
+        self._ophandle: _C._DispatchOperatorHandle = _C._dispatch_find_schema_or_throw(
+            name, overload_name
+        )
+        # Has the overload name, e.g. "foo.out" Computable from the schema but
+        # we cache here for convenience.
+        self._opname: str = str(schema.name)
+
+    @staticmethod
+    def define(namespaced_schema: str) -> "CustomOp":
+        r"""Creates a new CustomOp object.
+
+        In PyTorch, defining an op (short for "operator") is a two step-process:
+        - we need to define (create) the op
+        - we need to implement behavior for how the operator interacts with
+          various PyTorch subsystems, like CPU/CUDA Tensors, Autograd, etc.
+
+        This entrypoint defines the CustomOp object (the first step);
+        you must then perform the second step by calling various methods on
+        the CustomOp object.
+
+        Arguments:
+            namespaced_schema (str): The schema of the CustomOp.
+
+        Example::
+            >>> import numpy as np
+            >>>
+            >>> # Step 1: define the CustomOp
+            >>> numpy_sin = CustomOp.define('custom::numpy_sin')
+            >>>
+            >>> # Step 2: Register an implementation for various PyTorch subsystems
+            >>>
+            >>> # Register an implementation for CPU tensors
+            >>> @numpy_sin.impl('cpu'):
+            >>> def numpy_sin_impl_cpu(x):
+            >>>     return torch.from_numpy(np.sin(x.numpy()))
+            >>>
+            >>> # Register an implementation for CUDA tensors
+            >>> @numpy_sin.impl('cuda'):
+            >>> def numpy_sin_impl_cuda(x):
+            >>>     return torch.from_numpy(np.sin(x.cpu().numpy())).to(x.device)
+            >>>
+            >>> x = torch.randn(3)
+            >>> numpy_sin(x)  # calls numpy_sin_impl_cpu
+            >>>
+            >>> x_cuda = x.cuda()
+            >>> numpy_sin(x)  # calls numpy_sin_impl_cuda
+
+        """
+        ns, schema_str, schema = process_namespaced_schema_str(namespaced_schema)
+
+        lib = library.Library(ns, "FRAGMENT")
+        lib.define(schema_str)
+        result = CustomOp(lib, ns, schema, _private_access=True)
+
+        # NYI: autograd not supported
+        # In the near future we will either directly use the
+        # autograd_not_implemented kernels or make those the default fallback
+        # for the Autograd and ADInplaceOrView keys. Both of those are a bit tricky.
+        library.impl(lib, result._opname, "Autograd")(
+            get_autograd_not_implemented_kernel(weakref.proxy(result))
+        )
+
+        return result
+
+    def __call__(self, *args, **kwargs):
+        # Bypass torch.ops.* and directly do OperatorHandle::callBoxed.
+        # Using torch.ops.* is a bit of a pain (it can be slow and it has lifetime
+        # issues from caching operators that make testing CustomOp difficult).
+        result = _C._dispatch_call_boxed(self._ophandle, *args, **kwargs)
```

Comment:
Would this forbid us to access the overloaded op directly?

Reply:
Yes, the user is not allowed to access self._ophandle directly. Note that the user can still do torch.ops.custom.foo (or whatever) if they wanted, but I don't think there should be a reason for the user to do this